slm_lab/agent/algorithm/dqn.py

```diff
@@ -254,122 +253,103 @@ def post_body_init(self):
-        # print(batch_1)
-        # print("Batch 2: ")
-        # print(batch_2)
         # Package data into pytorch variables
-        float_data_list = [
+        float_data_names = [
             'states', 'actions', 'rewards', 'dones', 'next_states']
-        for k in float_data_list:
-            batch_1[k] = Variable(torch.from_numpy(batch_1[k]).float())
-            batch_2[k] = Variable(torch.from_numpy(batch_2[k]).float())
+        for ba, batch_b in enumerate(batches):
+            for k in float_data_names:
+                batch_b[k] = Variable(torch.from_numpy(batch_b[k]).float())
         # Concat state
         combined_states = torch.cat(
-            [batch_1['states'], batch_2['states']], dim=1)
+            [batch_b['states'] for batch_b in batches], dim=1)
         combined_next_states = torch.cat(
-            [batch_1['next_states'], batch_2['next_states']], dim=1)
+            [batch_b['next_states'] for batch_b in batches], dim=1)
         batch = {'states': combined_states,
                  'next_states': combined_next_states}
         # use recursive packaging to carry sub data
-        batch['sub_1'] = batch_1
-        batch['sub_2'] = batch_2
+        batch['batches'] = batches
         return batch
 
     def compute_q_target_values(self, batch):
-        batch_1 = batch['sub_1']
-        batch_2 = batch['sub_2']
-        # print("batch: {}".format(batch['states'].size()))
-        # print("batch: {}".format(batch['states']))
-        q_vals = self.net.wrap_eval(batch['states'])
-        # Use act_select network to select actions in next state
-        # Depending on the algorithm this is either the current
-        # net or target net
-        q_next_st_act_vals = self.online_net.wrap_eval(
+        batches = batch['batches']
+        # TODO make split eval tail-wise softmax, otherwise will skew to just one tail
+        q_sts = self.net.wrap_eval(batch['states'])
+        # TODO parametrize usage of eval or target_net
+        q_next_st_acts = self.online_net.wrap_eval(
             batch['next_states'])
 
-        # Select two sets of next actions
         # TODO Generalize to more than two tasks
-        _val, q_next_actions_1 = torch.max(
-            q_next_st_act_vals[:, :self.action_dims[0]], dim=1)
-        _val, q_next_actions_2 = torch.max(
-            q_next_st_act_vals[:, self.action_dims[0]:], dim=1)
+        start_idx = 0
+        q_next_acts = []
+        for body in self.agent.flat_nonan_body_a:
+            end_idx = start_idx + body.action_dim
+            _val, q_next_act_b = torch.max(
+                q_next_st_acts[:, start_idx:end_idx], dim=1)
+            q_next_acts.append(q_next_act_b)
+            start_idx = end_idx
         # Shift next actions_2 so they have the right indices
-        q_next_actions_2 = torch.add(q_next_actions_2, self.action_dims[0])
+        # TODO uhh what's this?
+        # q_next_actions_2 = torch.add(q_next_actions_2, self.action_dims[0])
```

Comment:
@lgraesser what is this and what could be the generalized version of the operation (refer to line 289-299)?